### PR TITLE
Move to SDK project

### PR DESCRIPTION
**Pull request by networkfusion**
## Description
* Moves project to SDK style.
* Adds .net48 build.
* Adds Fody.Costura to include dependent libs.
* Improves CI.
* Improves program to be testable.
* Removes warnings (such as use of `goto`).
* Adds some unit tests.
* Be less strict with package lock when building in VS, but enforce it in CI.
* Adds potential workaround for dependabot updates and package lock.

## Motivation and Context
* Fixes deprecated project type.
* Running the tool from the current nuget would fail due to dependent DLL's not being included.
* The project had many major warnings when running static analysis.
* There were no tests.

## How Has This Been Tested
* Multiple unit tests added
* Some integration tests added (compatible with CI)

## Screenshots
image
..
image
..
image

## Types of changes

- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [x] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [x] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [ ] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [ ] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [ ] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by josesimoes** [comment]
@networkfusion sorry for not looking into this sooner, but this one has to remain as it is in NET Framework. Otherwise it won't work with the VSCode extension which has to use Mono for building. This locks it to 4.7...

**Comment by networkfusion** [comment]
> @networkfusion sorry for not looking into this sooner, but this one has to remain as it is in NET Framework. Otherwise it won't work with the VSCode extension which has to use Mono for building. This locks it to 4.7...

I am ensuring that will still happen. Moving to SDK style projects should not stop this (as has happened with nanoff), but there are benefits (like not having to use old nuget installer) or older frameworks on win 10 & VS17. I am just starting to match the CI improvements with nanoff, which might help with the ability to build here.

**Comment by networkfusion** [comment]
interestingly, having an older FW does not throw like in VS:
image

Need to find why...

Also... using valid deployment files are not actually running (it seems, due to blank display on M5Core2.

File: nanoFrameworkDeployer/Program.cs [hunk 1]
```diff
@@ -10,227 +10,389 @@
...
 namespace nanoFrameworkDeployer
 {
     /// <summary>
-    /// This is the main program
+    /// This is the main program class
     /// </summary>
-    internal class Program
+    public class Program
     {
+        const int RETURN_CODE_SUCCESS = 0;
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:23]: in general, to follow our coding rules, please use ReturnCodeSuccess for enums, constants, classes, etc

File: nanoFrameworkDeployer/Program.cs [hunk 2]
```diff
@@ -10,227 +10,389 @@
...
+        /// <summary>
+        /// This allows overriding the default `System.IO` implementation using `System.IO.Abstractions`, 
+        /// so we can override it for tests.
+        /// </summary>
+        /// <remarks>
+        /// Usual filesystem operations start with `fileSystem` rather than just calling them directly.
+        /// The default calls `new FileSystem()` to load as per normal.
+        /// </remarks>
+        private static IFileSystem fileSystem = new FileSystem();
+
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:40]: extra line

File: nanoFrameworkDeployer/Program.cs [hunk 3]
```diff
@@ -10,227 +10,389 @@
...
+                deploymentFile.Close();
+                deploymentFile.Dispose();
+                return;
             }
-            else
+
+            ConsoleOutputHelper.Verbose("Known System serial ports include:");
+
+            //TODO: possibily match and/or discard entries when excluding or trying...
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:159]: missing space before the TODO

File: nanoFrameworkDeployer/Program.cs [hunk 4]
```diff
@@ -10,227 +10,389 @@
...
-        retryDebug:
-            bool connectResult = _device.DebugEngine.Connect(5000, true, true);
-            _message.Output($"Device connection result is: {connectResult}. Attempt {retryCount}/{numberOfRetries}");
+            ConnectDebugEngine(ref attemptCount, ref numberOfRetries);
 
-            if (!connectResult)
+            attemptCount = 1;
+
+            EraseDevice(ref attemptCount, ref numberOfRetries);
+
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:228]: extra line

File: nanoFrameworkDeployer/Program.cs [hunk 5]
```diff
@@ -10,227 +10,389 @@
...
+                    ConnectToDevice(ref attemptCount, ref numberOfRetries);
                 }
             }
+            else
+            {
+                ConsoleOutputHelper.Output($"Found: {_serialDebugClient.NanoFrameworkDevices.Count} devices");
+                return true;
+            }
+            return false;
+
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:266]: extra line

File: nanoFrameworkDeployer/Program.cs [hunk 6]
```diff
@@ -10,227 +10,389 @@
...
-                if (retryCount < numberOfRetries)
+                if (attemptCount <= numberOfRetries)
                 {
                     // Give it a bit of time
                     Thread.Sleep(400);
-                    retryCount++;
-                    goto retryErase;
+                    attemptCount++;
+                    EraseDevice(ref attemptCount, ref numberOfRetries);
+
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:288]: extra line

File: nanoFrameworkDeployer/Program.cs [hunk 7]
```diff
@@ -10,227 +10,389 @@
...
+        {
+            //FIXME: check file is formatted correctly?! delimiters might not be line breaks!
+            var ports = fileSystem.File.ReadAllLines(portExceptionFilePath);
+            if (ports.Length > 0)
+            {
+                var excludedPorts = new List<string>();
+                excludedPorts.AddRange(ports);
+                return excludedPorts;
+            }
+            return new List<string>();
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:333]: missing extra enter before the return

File: nanoFrameworkDeployer/Program.cs [hunk 8]
```diff
@@ -10,227 +10,389 @@
...
+                    EraseDevice(ref attemptCount, ref numberOfRetries);
+
+                }
+                else
+                {
+                    ConsoleOutputHelper.Error("ERROR: Could not erase, too many retries.");
+                    return false;
+                }
+            }
+            return true;
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:296]: missing extra enter before the return

File: nanoFrameworkDeployer/Program.cs [hunk 9]
```diff
@@ -10,227 +10,389 @@
...
+                    attemptCount++;
+                    ConnectToDevice(ref attemptCount, ref numberOfRetries);
                 }
             }
+            else
+            {
+                ConsoleOutputHelper.Output($"Found: {_serialDebugClient.NanoFrameworkDevices.Count} devices");
+                return true;
+            }
+            return false;
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/Program.cs:265]: missing extra enter before the return

File: nanoFrameworkDeployer/CommandLineOptions.cs [hunk 10]
```diff
@@ -32,18 +35,27 @@ public class CommandlineOptions
...
 
         /// <summary>
         /// Gets or sets the rebooting behavior after flash. True to reboot the device.
         /// </summary>
+        /// <remarks>
+        /// The default is false.
+        /// </remarks>
         [Option('r', "reboot", Required = false, HelpText = "Reboot the device after flash.")]
-        public bool RebootAfterFlash{ get; set; }
+        public bool RebootAfterFlash{ get; set; } = false;
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/CommandLineOptions.cs:51]: default is false anyway for a boolean, so why forcing it here?

File: nanoFrameworkDeployer/nanoFrameworkDeployer.csproj [hunk 11]
```diff
@@ -1,57 +1,68 @@
...
-    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
-    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
-    <ProjectGuid>{6AE83BF1-C167-45A6-A723-99FCADF2564E}</ProjectGuid>
     <OutputType>Exe</OutputType>
+    <ToolCommandName>nanoFrameworkDeployer</ToolCommandName>
+    <PackageOutputPath>../</PackageOutputPath>
+    <PackageId>nanoFrameworkDeployer</PackageId>
     <RootNamespace>nanoFrameworkDeployer</RootNamespace>
     <AssemblyName>nanoFrameworkDeployer</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
```
Comment (Ellerbach) [inline nanoFrameworkDeployer/nanoFrameworkDeployer.csproj:11]: I don't remember why but I'm quite sure there was a reason why this one is not a normal .NET 5.0/6.0 project... But anyway, makes more sense to move it to .NET 6.0

**Review by Ellerbach — COMMENTED** [review]
Couple of comments, nothing major!

File: nanoFrameworkDeployer/nanoFrameworkDeployer.csproj (hunk 11 above)
Comment (josesimoes) [reply nanoFrameworkDeployer/nanoFrameworkDeployer.csproj:11]: Debugger lib has to remain in 4.7.2 for as long as VS extensibility is locked to that version.
I recon that, with some effort we could have the build spitting both versions. Now considering that this is a **VERY** sensitive component of our toolbox I would stay way from messing with it until it's absolutely necessary.

File: nanoFrameworkDeployer/CommandLineOptions.cs (hunk 10 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/CommandLineOptions.cs]: Fixed.

File: nanoFrameworkDeployer/Program.cs (hunk 1 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs]: Fixed.

File: nanoFrameworkDeployer/Program.cs (hunk 3 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs]: Fixed.

File: nanoFrameworkDeployer/nanoFrameworkDeployer.csproj (hunk 11 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/nanoFrameworkDeployer.csproj:11]: I have some ideas about how we can do it... But will discuss internally on discord when I have the time to action it.

File: nanoFrameworkDeployer/Program.cs (hunk 7 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs:333]: Fixed.

File: nanoFrameworkDeployer/Program.cs (hunk 9 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs:264]: Fixed.

File: nanoFrameworkDeployer/Program.cs (hunk 8 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs:294]: Fixed.

File: nanoFrameworkDeployer/Program.cs (hunk 5 above)
Comment (networkfusion) [reply nanoFrameworkDeployer/Program.cs]: Fixed.

**Comment by networkfusion** [comment]
Should be ready for another review pass now (if needed).

**Pull request closed without merging**